lib/wibox/container/conditional.lua

```diff
@@ -0,0 +1,362 @@
+-- Note that it will be called only once for all templates.
+--
+-- The default value returns the `conditional` widget instance.
+--
+-- The callback has a single parameter (the widget instance).
+--
+-- @property argument_callback
+-- @param function
+
+function conditional:set_argument_callback(callback)
+    self._private.argument_callback = callback
+end
+
+function conditional:get_argument_callback()
+    return self._private.argument_callback
+end
+
+--- Reset this layout. The widget will be removed and the rotation reset.
+function conditional:reset()
+    self:set_widget(nil)
+end
+
+--- Evaluate all the `when` clauses and update the current widget(s).
+--
+-- @DOC_wibox_container_conditional_eval_EXAMPLE@
+--
+-- The signal is `property::selected_template` and has the following argument
+--    * The `wibox.container.conditional` widget instance
+--    * The current template
+--    * The old template
+--    * The current template widget instance
+--    * The old template widget instance
+--
+function conditional:eval()
+
+    -- Avoid called eval multiple timer per event loop cycle due to the
+    -- attached signals. This creates and destroy widgets for nothing
+    self._private.exec_once = false
+
+    local old_selected, selected = self._private.selected ,nil
+
+    for _, v in ipairs(self._private.templates or {}) do
+        if v.when then
+            if v.when(self._private.argument_callback(self)) then
+                selected = v
+                break
+            end
+        elseif self._private.default_when then
+            local ret = self._private.default_when(v, self._private.argument_callback(self))
+
+            -- The function must return something or is considered broken
+            assert(type(ret) == "boolean", "The conditional layout `default_when`"..
+                " must return a boolean"
+            )
+
+            if ret then
+                selected = v
+                break
+            end
+        else
+            assert(false, "A conditional template has neither a `when` property"..
+                " nor a default_when"
+            )
+        end
+    end
+
+    -- Nothing changed, move on
+    if selected == self._private.selected then
+        if selected and selected.updated_callback then
+            selected.updated_callback(self._private.argument_callback())
+        end
+
+        return
+    end
+
+    if self._private.selected and self._private.selected.deselected_callback then
+        self._private.selected:deselected_callback()
+    end
+
+    self._private.selected = selected
+
+    if self._private.selected and self._private.selected.selected_callback then
+        self._private.selected:selected_callback()
+    end
+
+    local old_widget = self._private.widget
+
+    -- Create the widget unless it's already one.
+    self._private.widget = selected and
+        base.make_widget_from_value(selected) or nil
```

Comment:
So that's the place that does this... why? Why not simply create all widgets upfront?

Reply:
Answered above.